src/pybind/mgr/orchestrator.py

```diff
@@ -845,19 +903,34 @@ def _oremote(self, meth, args, kwargs):
         self.log.debug("_oremote {} -> {}.{}(*{}, **{})".format(self.module_name, o, meth, args, kwargs))
         return self.remote(o, meth, *args, **kwargs)
 
-    def _orchestrator_wait(self, completions):
+    def _orchestrator_wait(self, completions, raise_if_errored=True):
         # type: (List[_Completion]) -> None
         """
-        Helper to wait for completions to complete (reads) or
+        Wait for completions to complete (reads) or
         become persistent (writes).
 
         Waits for writes to be *persistent* but not *effective*.
+
+        :param completions: List of Completions
+        :param raise_if_errored:
+        :raises OrchestratorError: A completion failed. Only if ``raise_if_errored == True``
+        :raises NoOrchestrator:
+        :raises ImportError: no `orchestrator_cli` module or backend not found.
         """
         while not self.wait(completions):
             if any(c.should_wait for c in completions):
                 time.sleep(5)
             else:
                 break
 
-        if all(hasattr(c, 'error') and getattr(c, 'error') for c in completions):
-            raise Exception([getattr(c, 'error') for c in completions])
+        if raise_if_errored and completions and any(c.is_errored for c in completions):
+            if len(completions) == 1 and completions[0].exception:
+                # completions[0].exception is an exception from a module from an other
+                # sub-interpreter. Which is ont catchable here in this sub-interpreter.
+                e = completions[0].exception
+                cls_name = type(e).__name__
+                cls = getattr(sys.modules[__name__], cls_name, Exception)
+                if cls.__name__ == cls_name:
+                    raise cls(str(e))
+                raise e  # If not a an error of the standard lib, it's not going to be catchable
+            raise OrchestratorError([c.exception if c.exception else c.result for c in completions])
```

Comment:
Why only the first completion of the list is in another sub-interpreter?

Reply:
This is just the hand-wavy case. What would you raise, if you're getting three completions, with

1. first one being a `ZeroDivisionError`
2. second being a successful completion
3. third one being a `OrchestratorValidationError`

Reply:
None. 
I think that the wait method should be responsible only of checking the progress of the completions objects. If we have an error in one completion,  this completion should be "completed" ( filling properly the exception/is_errored properties) and therefore excluded of the list of completions processed by the wait method.

Reply:
Hmm. Even though `wait` accepts several completions at once, I assumed that each completion itself was independent. If that is actually the case, then it would seem beneficial to be able to report multiple failures / successes. from this method. Or raising no exception and handling reporting some other way. Though, perhaps I'm missing something.

Reply:
Here is a `ceph orchestrator wait` implementation that waits for multiple completions: 

https://github.com/sebastian-philipp/ceph/commit/663f6c6f11e437d8deea91b1c1db019e92884127#diff-b3018215c5634fd57e880aaec0a79dcaR85

Reply:
First of all... awesome utilization of generators and decorators  in your _ceph orchestrator wait_ implementation!!! Congratulations!!

Then two different things:
**Modification in orchestrator.py:**
I continue thinking that the utility of the _orchestrator_wait method is what appears in the docstring: 
"Wait for completions to complete (reads) or become persistent (writes)."

I do not thing that will be useful to try to manage errors for each completion(or only for part of them) here. Even i think that probably we would remove the previous completions error  management in this function.
In my opinion each completion should be independent, and has attributes to allow the consumer to know what is the state of the completion object, get the error or get the result.

So for me, completions error management in this method is not needed, and probalby doing that we are moving form what was the initial design. ( this is not bad... but in this case i do not think that we are going to obtain advantages over this initial design)

**Implementation that waits for multiple completions in orchestrator_cli:**
This is a modification in the orchestrator-cli and i think that here we are talking more about @noahdesu comments.
Is ok... but what is the use case of this method? orchestrator-cli is a command line tool, this kind of tools are frequently used in a sequential and interactive way.
I can not see too much utility of this in the orchestrator_cli. However the implementation you has made is awesome for other orchestrators clients ,  for example for the dasboard, where  is more easy (and even expected) to have several operations running at the same time.

Reply:
@jmolmo & @noahdesu : Yeah, let's remove this from `_orchestrator_wait()` altogether. If we cannot properly do error handling here in this method, then we shouldn't try to do so. 

> several operations running at the same time.

The use case is also easily applicable to the cli:

``` bash
for host in $cluster_hosts ; do
  ceph orchestrator osd create $host /dev/sda:/dev/sdb --wait=false  # schedules some completions
done
ceph orchestrator wait  # waits for all of them.
```

Reply:
You are right... go ahead also with the change in the orchestrator_cli ... is is an excellent example of how to do things very well using orchestrator  features. :-)

Reply:
fixed.